sdv/datasets/demo.py

```diff
@@ -39,41 +39,153 @@ def _validate_output_folder(output_folder_name):
-        raise ValueError(
-            f"Invalid dataset name '{dataset_name}'. "
-            'Make sure you have the correct modality for the dataset name or '
-            "use 'get_available_demos' to get a list of demo datasets."
+def _list_objects(prefix):
+    """List all objects under a given prefix using pagination.
+
+    Args:
+        prefix (str):
+            The S3 prefix to list.
+
+    Returns:
+        list[dict]:
+            A list of object summaries.
+    """
+    client = _create_s3_client()
+    contents = []
+    paginator = client.get_paginator('list_objects_v2')
+    for resp in paginator.paginate(Bucket=BUCKET, Prefix=prefix):
+        contents.extend(resp.get('Contents', []))
+
+    if not contents:
+        raise DemoResourceNotFoundError(f"No objects found under '{prefix}' in bucket '{BUCKET}'.")
+
+    return contents
+
+
+def _search_contents_keys(contents, match_fn):
+    """Return list of keys from ``contents`` that satisfy ``match_fn``.
+
+    Args:
+        contents (list[dict]):
+            S3 list_objects-like contents entries.
+        match_fn (callable):
+            Function that receives a key (str) and returns True if it matches.
+
+    Returns:
+        list[str]:
+            Keys in their original order that matched the predicate.
+    """
+    matches = []
+    for entry in contents or []:
+        key = entry.get('Key', '')
+        try:
+            if match_fn(key):
+                matches.append(key)
+        except Exception:
+            continue
+
+    return matches
+
+
+def _find_data_zip_key(contents, dataset_prefix):
+    """Find the 'data.zip' object key under dataset prefix, case-insensitive.
+
+    Args:
+        contents (list[dict]):
+            List of objects from S3.
+        dataset_prefix (str):
+            Prefix like 'single_table/dataset/'.
+
+    Returns:
+        str:
+            The key to the data zip if found.
+    """
+    prefix_lower = dataset_prefix.lower()
+
+    def is_data_zip(key):
+        return key.lower() == f'{prefix_lower}data.zip'
+
+    matches = _search_contents_keys(contents, is_data_zip)
+    if matches:
+        return matches[0]
+
+    raise DemoResourceNotFoundError("Could not find 'data.zip' for the requested dataset.")
+
+
+def _get_first_v1_metadata_bytes(contents, dataset_prefix):
+    """Find and return bytes of the first V1 metadata JSON under `dataset_prefix`.
+
+    Scans S3 listing `contents` and, for any JSON file directly under the dataset prefix,
+    downloads and returns its bytes if it contains METADATA_SPEC_VERSION == 'V1'.
+
+    Returns:
+        bytes:
+            The bytes of the first V1 metadata JSON.
+    """
+    prefix_lower = dataset_prefix.lower()
+
+    def is_direct_json_under_prefix(key):
+        key_lower = key.lower()
+        return (
+            key_lower.startswith(prefix_lower)
+            and key_lower.endswith('.json')
+            and key_lower.count('/') == prefix_lower.count('/')
```

Comment:
should we check that metadata is also in the name? What if we add a different json in the future